File: CometClient/Assets/Code/Networking/Packets.cs
using System;


namespace net
{

	//In C# the effect of 'using' aliases is limited to file scope.
	using uint8_t = Byte;
	using uint16_t = UInt16;
	using uint32_t = UInt32;
	using entity_id = UInt32; //TODO: Find some way to emulate global typedefs.

	enum packet_type { client_input, server_state, shape_request, shape_description };

	interface BinarySerializable
	{
		int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index);
	}

	class Header : BinarySerializable
	{
		public uint8_t protocol_id;
		public uint32_t sequence_number;
		public uint8_t packet_type;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref protocol_id);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref sequence_number);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref packet_type);
			return current_index - start_index;
		}
	};

	class ServerHeader : BinarySerializable
	{
		public Header common_header;
		public uint32_t ack;
		public uint32_t ack_bitfield;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += common_header.Process(io_mode, packet_data, current_index);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref ack);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref ack_bitfield);
			return current_index - start_index;
		}
	};

	class ClientInputPayload : BinarySerializable
	{
		public entity_id entity_id;
		public uint8_t duration;
		public uint16_t count; //count=0 will be used as client heartbeat.
		public uint8_t[] inputs;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref entity_id);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref duration);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref count);
			for (int i = 0; i < count; i++)
			{
				current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref inputs[i]);
			}
			return current_index - start_index;
		}
	};

	class ServerObject : BinarySerializable
	{
		public entity_id entity_id; //TODO: Originally this was called 'type'. Find out why.
		public double phi;
		public double x;
		public double y;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref entity_id);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref phi);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref x);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref y);
			return current_index - start_index;
		}
	};

	class ServerStatePayload : BinarySerializable
	{
		public uint16_t count;
		public ServerObject[] objects;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref count);
			for (int i = 0; i < count; i++)
			{
				current_index += objects[i].Process(io_mode, packet_data, current_index);
			}
			return current_index - start_index;
		}
	};

	class ShapeRequest : BinarySerializable
	{
		public entity_id entity_id;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref entity_id);
			return current_index - start_index;
		}
	};

	class ShapeDescription : BinarySerializable
	{
		public uint16_t vertex_count;
		public uint16_t triangle_count;
		public float[] vertices;
		public float[] uvs;
		public uint16_t[] triangles;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref vertex_count);
			current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref triangle_count);
			for (int i = 0; i < vertex_count * 2; i++)
			{
				current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref vertices[i]);
			}
			for (int i = 0; i < vertex_count * 2; i++)
			{
				current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref uvs[i]);
			}
			for (int i = 0; i < triangle_count * 3; i++)
			{
				current_index += net.BinarySerializer.Process(io_mode, packet_data, current_index, ref triangles[i]);
			}
			return current_index - start_index;
		}
	};

	class Packet<H, P> where H : BinarySerializable where P : BinarySerializable
	{
		public H header;
		public P payload;

		public int Process(net.BinarySerializer.IOMode io_mode, Byte[] packet_data, int start_index)
		{
			int current_index = start_index;
			current_index += header.Process(io_mode, packet_data, current_index);
			current_index += payload.Process(io_mode, packet_data, current_index);
			return current_index - start_index;
		}
	};

}

File: CometServer/Networking/Packets.cpp
#include "Packets.hpp"
#include "BinarySerialization.hpp"


namespace net
{

	bool Header::operator==(const Header& other) const
	{
		return protocol_id == other.protocol_id && sequence_number == other.sequence_number && packet_type == other.packet_type;
	}

	template<typename io_mode>
	size_t Header::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += io_mode::Process(packet_data_current, protocol_id);
		packet_data_current += io_mode::Process(packet_data_current, sequence_number);
		packet_data_current += io_mode::Process(packet_data_current, packet_type);
		return packet_data_current - packet_data_start;
	}

	bool ServerHeader::operator==(const ServerHeader& other) const
	{
		return common_header == other.common_header && ack == other.ack && ack_bitfield == other.ack_bitfield;
	}

	template<typename io_mode>
	size_t ServerHeader::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += common_header.IO<io_mode>(packet_data_current);
		packet_data_current += io_mode::Process(packet_data_current, ack);
		packet_data_current += io_mode::Process(packet_data_current, ack_bitfield);
		return packet_data_current - packet_data_start;
	}

	void ClientInputPayload::DeepCopyFrom(const ClientInputPayload& that)
	{
		uint8_t* these_inputs = this->inputs;
		*this = that;
		this->inputs = these_inputs;
		std::memcpy(this->inputs, that.inputs, that.count);
	}

	bool ClientInputPayload::operator==(const ClientInputPayload& other) const
	{
		if (entity_id != other.entity_id) return false;
		if (duration != other.duration) return false;
		if (count != other.count) return false;
		for (int i = 0; i < count; ++i)
		{
			if (inputs[i] != other.inputs[i]) return false;
		}
		return true;
	}

	template<typename io_mode>
	size_t ClientInputPayload::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += io_mode::Process(packet_data_current, entity_id);
		packet_data_current += io_mode::Process(packet_data_current, duration);
		packet_data_current += io_mode::Process(packet_data_current, count);
		for (uint8_t* i = inputs; i - inputs < count; ++i)
		{
			packet_data_current += io_mode::Process(packet_data_current, *i);
		}
		return packet_data_current - packet_data_start;
	}

	bool ServerObject::operator==(const ServerObject& other) const
	{
		return entity_id == other.entity_id && phi == other.phi && x == other.x && y == other.y;
	}

	template<typename io_mode>
	size_t ServerObject::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += io_mode::Process(packet_data_current, entity_id);
		packet_data_current += io_mode::Process(packet_data_current, phi);
		packet_data_current += io_mode::Process(packet_data_current, x);
		packet_data_current += io_mode::Process(packet_data_current, y);
		return packet_data_current - packet_data_start;
	}

	void ServerStatePayload::DeepCopyFrom(const ServerStatePayload& that)
	{
		ServerObject* these_objects = this->objects;
		*this = that;
		this->objects = these_objects;
		std::memcpy(this->objects, that.objects, that.count);
	}

	bool ServerStatePayload::operator==(const ServerStatePayload& other) const
	{
		if (!(count == other.count)) return false;
		for (int i = 0; i < count; ++i)
		{
			if (!(objects[i] == other.objects[i])) return false;
		}
		return true;
	}

	template<typename io_mode>
	size_t ServerStatePayload::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += io_mode::Process(packet_data_current, count);
		for (ServerObject* i = objects; i - objects < count; ++i) //TODO: Somehow handle overindexing.
		{
			packet_data_current += i->IO<io_mode>(packet_data_current);
		}
		return packet_data_current - packet_data_start;
	}

	bool ShapeRequestPayload::operator==(const ShapeRequestPayload& other) const
	{
		return entity_id == other.entity_id;
	}

	template<typename io_mode>
	size_t ShapeRequestPayload::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += io_mode::Process(packet_data_current, entity_id);
		return packet_data_current - packet_data_start;
	}

	void ShapeDescriptionPayload::DeepCopyFrom(const ShapeDescriptionPayload& that)
	{
		this->entity_id = that.entity_id;
		this->vertex_count = that.vertex_count;
		this->triangle_count = that.triangle_count;
		std::memcpy(this->vertices, that.vertices, that.vertex_count);
		std::memcpy(this->uvs, that.uvs, that.vertex_count);
		std::memcpy(this->triangles, that.triangles, that.triangle_count);
	}

	bool ShapeDescriptionPayload::operator==(const ShapeDescriptionPayload& other) const
	{
		if (!(entity_id == other.entity_id)) return false;
		if (!(vertex_count == other.vertex_count)) return false;
		if (!(triangle_count == other.triangle_count)) return false;
		for (int i = 0; i < vertex_count * 2; ++i)
		{
			if (!(vertices[i] == other.vertices[i])) return false;
		}
		for (int i = 0; i < vertex_count * 2; ++i)
		{
			if (!(uvs[i] == other.uvs[i])) return false;
		}
		for (int i = 0; i < triangle_count * 3; ++i)
		{
			if (!(triangles[i] == other.triangles[i])) return false;
		}
		return true;
	}

	template<typename io_mode>
	size_t ShapeDescriptionPayload::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += io_mode::Process(packet_data_current, entity_id);
		packet_data_current += io_mode::Process(packet_data_current, vertex_count);
		packet_data_current += io_mode::Process(packet_data_current, triangle_count);
		for (float* i = vertices; i - vertices < vertex_count * 2; ++i) //TODO: Somehow handle overindexing.
		{
			packet_data_current += io_mode::Process(packet_data_current, *i);
		}
		for (float* i = uvs; i - uvs < vertex_count * 2; ++i) //TODO: Somehow handle overindexing.
		{
			packet_data_current += io_mode::Process(packet_data_current, *i);
		}
		for (uint16_t* i = triangles; i - triangles < triangle_count * 3; ++i) //TODO: Somehow handle overindexing.
		{
			packet_data_current += io_mode::Process(packet_data_current, *i);
		}
		return packet_data_current - packet_data_start;
	}

	template<typename H, typename P>
	bool Packet<H, P>::operator==(const Packet<H, P>& other) const
	{
		return header == other.header && payload == other.payload;
	}

	template<typename H, typename P>
	template<typename io_mode>
	size_t Packet<H, P>::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += header.IO<io_mode>(packet_data_current);
		packet_data_current += payload.IO<io_mode>(packet_data_current);
		return packet_data_current - packet_data_start;
	}

	template<typename H, typename P>
	bool PointeredPacket<H, P>::operator==(const PointeredPacket<H, P>& other) const
	{
		return *header == *(other.header) && *payload == *(other.payload);
	}

	template<typename H, typename P>
	template<typename io_mode>
	size_t PointeredPacket<H, P>::IO(uint8_t* packet_data_start)
	{
		uint8_t* packet_data_current = packet_data_start;
		packet_data_current += header->IO<io_mode>(packet_data_current);
		packet_data_current += payload->IO<io_mode>(packet_data_current);
		return packet_data_current - packet_data_start;
	}

	AbstractExportStrategy::AbstractExportStrategy() {};

	AbstractExportStrategy::~AbstractExportStrategy() {};

	AbstractImportStrategy::AbstractImportStrategy() {};

	AbstractImportStrategy::~AbstractImportStrategy() {};

	//Explicit instantiations.
	template size_t Header::IO<Read>(uint8_t*);
	template size_t Header::IO<Write>(uint8_t*);
	template size_t ServerHeader::IO<Read>(uint8_t*);
	template size_t ServerHeader::IO<Write>(uint8_t*);
	template size_t ClientInputPayload::IO<Read>(uint8_t*);
	template size_t ClientInputPayload::IO<Write>(uint8_t*);
	template size_t ServerObject::IO<Read>(uint8_t*);
	template size_t ServerObject::IO<Write>(uint8_t*);
	template size_t ServerStatePayload::IO<Read>(uint8_t*);
	template size_t ServerStatePayload::IO<Write>(uint8_t*);
	template size_t ShapeRequestPayload::IO<Read>(uint8_t*);
	template size_t ShapeRequestPayload::IO<Write>(uint8_t*);
	template size_t ShapeDescriptionPayload::IO<Read>(uint8_t*);
	template size_t ShapeDescriptionPayload::IO<Write>(uint8_t*);
	template class Packet<Header, ClientInputPacket>;
	template size_t Packet<Header, ClientInputPacket>::IO<Read>(uint8_t*);
	template size_t Packet<Header, ClientInputPacket>::IO<Write>(uint8_t*);
	template class Packet<ServerHeader, ServerStatePayload>;
	template size_t Packet<ServerHeader, ServerStatePayload>::IO<Read>(uint8_t*);
	template size_t Packet<ServerHeader, ServerStatePayload>::IO<Write>(uint8_t*);
	template class Packet<Header, ShapeRequestPayload>;
	template size_t Packet<Header, ShapeRequestPayload>::IO<Read>(uint8_t*);
	template size_t Packet<Header, ShapeRequestPayload>::IO<Write>(uint8_t*);
	template class Packet<ServerHeader, ShapeDescriptionPayload>;
	template size_t Packet<ServerHeader, ShapeDescriptionPayload>::IO<Read>(uint8_t*);
	template size_t Packet<ServerHeader, ShapeDescriptionPayload>::IO<Write>(uint8_t*);
	template class PointeredPacket<Header, ClientInputPacket>;
	template size_t PointeredPacket<Header, ClientInputPacket>::IO<Read>(uint8_t*);
	template size_t PointeredPacket<Header, ClientInputPacket>::IO<Write>(uint8_t*);
	template class PointeredPacket<ServerHeader, ServerStatePayload>;
	template size_t PointeredPacket<ServerHeader, ServerStatePayload>::IO<Read>(uint8_t*);
	template size_t PointeredPacket<ServerHeader, ServerStatePayload>::IO<Write>(uint8_t*);
	template class PointeredPacket<Header, ShapeRequestPayload>;
	template size_t PointeredPacket<Header, ShapeRequestPayload>::IO<Read>(uint8_t*);
	template size_t PointeredPacket<Header, ShapeRequestPayload>::IO<Write>(uint8_t*);
	template class PointeredPacket<ServerHeader, ShapeDescriptionPayload>;
	template size_t PointeredPacket<ServerHeader, ShapeDescriptionPayload>::IO<Read>(uint8_t*);
	template size_t PointeredPacket<ServerHeader, ShapeDescriptionPayload>::IO<Write>(uint8_t*);

}
